### PR TITLE
 Added OpenLDAP database for ACME

### PR DESCRIPTION
**Pull request by edewata**
A new `OpenLDAPDatabase` class has been added to support OpenLDAP database for ACME. The class is currently identical to the existing `LDAPDatabase` class, but it's provided for implementing OpenLDAP-specific code in the future.

A new schema file, a sample initialization file, and a sample configuration file for OpenLDAP have been added.

The `pki-server acme-database-*` commands have also been modified to support an `openldap` type.

The ACME database documentation has been updated as well:
https://github.com/edewata/pki/blob/acme-openldap/docs/installation/acme/Configuring_ACME_Database.md

Changes to `LDAPDatabase` class:
* Previously the certificate data for ACME certificates was stored as `userCertificate=<data>` in DS, which doesn't seem to work in OpenLDAP. To support both LDAP servers, the certificate data is now stored as `userCertificate;binary=<data>`.

* Previously the identifiers for ACME order and authorization records were stored as `acmeIdentifier;<type>=<value>` in DS, which doesn't seem to be supported in OpenLDAP. To support both LDAP servers, the identifiers are now stored as `acmeIdentifier=<type>:<value>`.

### CODE REVIEW

File: base/server/python/pki/server/cli/acme.py [hunk 1]
```diff
@@ -607,7 +608,7 @@ def execute(self, argv):
         database_type = DATABASE_TYPES.get(database_class)
         print('  Database Type: %s' % database_type)
 
-        if database_type == 'ldap':
+        if database_type == 'ldap' or database_type == 'openldap':
```
Comment (cipherboy) [inline base/server/python/pki/server/cli/acme.py:611]: Should we split into three instead of two?

`ldap` -- generic base, `389ds` and `openldap`? Then if there are any overtly 389ds-specific things, we can do a "general" implementation that might work on LDAPMcLDAP's implementation, and then overwrite it with 389-ds specific one too? Comment kinda applies everywhere, not just here. 

i.e., make it explicit that `ldap` really means 389ds? Then making openldap seems more natural.

But up to you.

Comment (edewata) [reply base/server/python/pki/server/cli/acme.py:611]: Actually I was considering to rename `ldap` to `ds` after this PR, but it needs some discussions:
* `389ds` refers to the upstream project. Downstream it's `rhds`. So should we use `ds` instead? The class name will be`DSDatabase`.
* I'm not sure if we want to keep the generic `ldap` option since there's no product that can be used with it, but we can keep the `LDAPDatabase` as the base class.
* IPA will need to be updated as well.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java [hunk 2]
```diff
@@ -432,13 +430,14 @@ private static ACMEOrder loadOrder(LDAPEntry entry)
...
-            attr = entry.getAttribute(ATTR_IDENTIFIER + ";" + type);
-            for (String value : Collections.list(attr.getStringValues())) {
-                identifiers.add(new ACMEIdentifier(type, value));
-            }
+        attr = entry.getAttribute(ATTR_IDENTIFIER);
+        for (String identifier : attr.getStringValueArray()) {
+            String[] parts = StringUtils.split(identifier, ":", 2);
+            String type = parts[0];
+            String value = parts[1];
+            identifiers.add(new ACMEIdentifier(type, value));
```
Comment (frasertweedale) [inline base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:438]: Ad-hoc attribute subtypes via ';' is standard LDAP stuff; it's disappointing OpenLDAP apparently doesn't support it properly.  And this is a breaking change for existing databases.  So could we put this behavoiur in the `OpenLDAPDatabase` class only, and leave `LDAPDatabase` as-is?  (This comment also applies at several other place in the diff).

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java [hunk 3]
```diff
@@ -432,13 +430,14 @@ private static ACMEOrder loadOrder(LDAPEntry entry)
...
         }
-        order.setIdentifiers(identifiers.toArray(new ACMEIdentifier[0]));
+        order.setIdentifiers(identifiers.toArray(new ACMEIdentifier[identifiers.size()]));
```
Comment (frasertweedale) [inline base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:440]: Why this change?  The argument is only needed to fix the return type parameter, and it was written in the way suggested in the official doc: https://docs.oracle.com/javase/7/docs/api/java/util/List.html#toArray(T[]) .

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java [hunk 4]
```diff
@@ -837,7 +834,7 @@ public ACMECertificate getCertificate(String certID) throws Exception {
         LDAPAttribute attr = entry.getAttribute(ATTR_CREATED);
         certificate.setCreationTime(dateFormat.parse(attr.getStringValues().nextElement()));
 
-        attr = entry.getAttribute(ATTR_USER_CERTIFICATE);
+        attr = entry.getAttribute(ATTR_USER_CERTIFICATE + ";binary");
```
Comment (frasertweedale) [inline base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:837]: This change is fine to make in `LDAPDatabase`.  But maybe we should put the `";binary"` suffix in the definition of `ATTR_USER_CERTIFICATE`.  The current code adds the suffix for every use of the attribute (as it should) but putting it in `ATTR_USER_CERTIFICATE` should ensure that any new uses of the attribute includes the suffix (and eliminates a duplication).

**Review by frasertweedale — CHANGES_REQUESTED** [review]
A few suggestions.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 2 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:438]: I think we should minimize the difference between LDAP implementations to make it easier to migrate from one database to another. The ACME is a tech preview in PKI 10.8 and 10.9 and we've already [warned](https://github.com/dogtagpki/pki/blob/master/docs/installation/acme/Installing_PKI_ACME_Responder.md) about possible incompatible changes. Note that we don't have any automated database upgrade mechanism, so it's better to fix the schema before ACME becomes officially supported. Also, the ACME database is supposed to be transient anyway, so people can just remove their existing ACME subtree.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 3 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:440]: OK, in the past the one with 0-size array was less efficient, but apparently that's not the case anymore:
https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size
I'll undo this change.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 4 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:837]: OK, we can do that.

File: base/server/python/pki/server/cli/acme.py (hunk 1 above)
Comment (edewata) [reply base/server/python/pki/server/cli/acme.py]: I've added the DS database. The doc has been updated as well.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 3 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java]: Removed.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 4 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java]: Done.

**Comment by edewata** [comment]
Please take a look at the updated PR. Thanks!

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 2 above)
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:438]: After further research, I agree with changing it.  Although the LDAP data model allows arbitrary attribute description options (tags), they are supposed to be registered with IANA and an implementation SHALL treat an attribute description with an unrecognised attribute option as unrecognised.  ( https://tools.ietf.org/html/rfc4512#section-2.5 ).

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java [hunk 5]
```diff
@@ -432,11 +430,12 @@ private static ACMEOrder loadOrder(LDAPEntry entry)
...
-        for (String type : IDENTIFIER_TYPES) {
-            attr = entry.getAttribute(ATTR_IDENTIFIER + ";" + type);
-            for (String value : Collections.list(attr.getStringValues())) {
-                identifiers.add(new ACMEIdentifier(type, value));
-            }
+        attr = entry.getAttribute(ATTR_IDENTIFIER);
+        for (String identifier : attr.getStringValueArray()) {
+            String[] parts = StringUtils.split(identifier, ":", 2);
+            String type = parts[0];
+            String value = parts[1];
```
Comment (frasertweedale) [inline base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:437]: Here let us handle the case where ':' is not present in the string.  Both to handle old data, but more importantly to be robust in case of incorrect/corrupt data in database somehow.  Better to handle gracefully (meaningful error or perhaps just skip the value) than fail with index-out-of-bounds exception.

Whether we provide a fallback to handle the original scheme that used attribute description option I don't mind either way, but it would be nice.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java [hunk 6]
```diff
@@ -564,14 +563,12 @@ private ACMEAuthorization getAuthorization(String authzID, LoadChallenges clc)
...
-                authz.setIdentifier(
-                    new ACMEIdentifier(type, attr.getStringValues().nextElement()));
-                break;
-            }
-        }
+        attr = entry.getAttribute(ATTR_IDENTIFIER);
+        String identifier = attr.getStringValues().nextElement();
+        String[] parts = StringUtils.split(identifier, ":", 2);
+        String type = parts[0];
+        String value = parts[1];
```
Comment (frasertweedale) [inline base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:570]: Here too we should handle `parts.length < 2`.

**Review by frasertweedale — CHANGES_REQUESTED** [review]
One more request to make the processing more robust against old or invalid indentifier data.  Besides that it LGTM (but I haven't tested).

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 5 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:440]: OK, I added a code to throw an exception if the identifier has an incorrect format, but I think this code will only read `acmeIdentifier`, not `acmeIdentifier;dns`. If we want to read the old attribute we probably would have to keep the old code.

I think going forward we should focus on upgrading the database instead of getting stuck with an old database. There's a plan to implement a [database upgrade framework](https://pagure.io/dogtagpki/issue/710) in PKI 10.10, but I don't know how much will actually get implemented. Hopefully we can use that to migrate ACME database.

File: base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java (hunk 6 above)
Comment (edewata) [reply base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java:579]: Same as above.

**Comment by edewata** [comment]
Thanks for the reviews! Merged.
